### PR TITLE
Initial addition of dataset access module

### LINKED ISSUE #12: Control dataset access
All users that need to consume the data set need at minimum dataset viewer. Would best to have an array of users and be able to access.

### PR DESCRIPTION
**Pull request by milesmatthias**
Fixes #12.

### CODE REVIEW

**Comment by aaron-lane** [comment]
Would this fix #12?

File: examples/dataset_access/main.tf [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+
+module "bigquery" {
+  source            = "../.."
+  dataset_id        = "plane"
+  dataset_name      = "plane"
+  description       = "some description"
+  expiration        = var.expiration
+  project_id        = var.project_id
+  location          = "US"
+  tables            = var.tables
```
Comment (morgante) [inline examples/dataset_access/main.tf]: Tests/examples should not be excessively variabilized. Let's just provide pre-built example tables.

File: examples/dataset_access/main.tf [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+  dataset_labels    = var.dataset_labels
+}
+
+module "dataset_access" {
+  source     = "../../modules/dataset_access"
+  dataset_id = module.bigquery.dataset_id
+  project    = var.project_id
+
+  preserve_special_groups = true
+  roles_json              = var.roles_json
```
Comment (morgante) [inline examples/dataset_access/main.tf]: Don't make this a variable: instead let's provide an example of what this actually looks like.

**Comment by milesmatthias** [comment]
I'm going to need some help setting up the CI IAM groups & admin email to get these CI tests passing. Not sure who the best person is to talk to about that.

**Comment by milesmatthias** [comment]
Need https://github.com/GoogleCloudPlatform/cloud-foundation-toolkit/pull/509 merged to add directory of binaries to ignore linting.

**Comment by morgante** [comment]
@milesmatthias Merged the new devtools image.

**Comment by milesmatthias** [comment]
Running into https://github.com/terraform-google-modules/terraform-google-bigquery/issues/26#issuecomment-568281951, so either need to remove some of the clustering support, or update this feature to require Google v3.0.0 provider.

**Comment by milesmatthias** [comment]
Linting will be fixed by https://github.com/GoogleCloudPlatform/cloud-foundation-toolkit/pull/523.

**Comment by milesmatthias** [comment]
@morgante would love any thoughts you have here to help unblock me. Thanks!

**Comment by milesmatthias** [comment]
@morgante this looks good to go

File: modules/dataset_access/Makefile [hunk 3]
```diff
@@ -0,0 +1,53 @@
+# Copyright 2019 Google LLC
```
Comment (morgante) [inline modules/dataset_access/Makefile]: I don't think we should have a module-specific Makefile. Instead, could we fold any special commands into the standard module Makefile?

Possibly with custom helpers: https://github.com/terraform-google-modules/terraform-google-kubernetes-engine/blob/master/test/task_helper_functions.sh

**Comment by mikhail-khodorovskiy** [comment]
We are very interested in this capability to be merged.  Is there any ETA on that?

**Comment by 06kellyjac** [comment]
We also need access control (and encryption). :)

Out of interest, why is this done by executing a binary vs setting the [`access` block](https://www.terraform.io/docs/providers/google/r/bigquery_dataset.html#access)?

**Comment by umairidris** [comment]
@morgante this can be closed

**Comment by umairidris** [comment]
This was fixed by #48.

**Pull request closed without merging**